src/normalizer/DecompFile.zig

```diff
@@ -133,14 +246,132 @@ pub fn compressToFile(self: *DecompFile, filename: []const u8) !void {
+            var fits_2 = true;
+            var fits_8 = true;
+            for (diff.value.seq[0..entry.value.len]) |s| {
+                if (s >= (1 << 2)) {
+                    fits_2 = false;
+                }
+                if (s >= (1 << 8)) {
+                    fits_8 = false;
+                }
+            }
+            if (fits_2) {
+                try out.writeBits(@enumToInt(Opcode.increment_key_3_and_set_value_seq_0_2bit_and_emit), @bitSizeOf(Opcode));
+                try out.writeBits(diff.value.seq[0], 2);
+            } else if (fits_8) {
+                try out.writeBits(@enumToInt(Opcode.increment_key_3_and_set_value_seq_0_8bit_and_emit), @bitSizeOf(Opcode));
+                try out.writeBits(diff.value.seq[0], 8);
+            } else {
+                try out.writeBits(@enumToInt(Opcode.increment_key_3_and_set_value_seq_0_21bit_and_emit), @bitSizeOf(Opcode));
+                try out.writeBits(diff.value.seq[0], 21);
+            }
+        } else if (seq_0_1_change and mem.eql(u8, &diff.key, &[4]u8{ 0, 1, 0, 0 })) {
+            try out.writeBits(@enumToInt(Opcode.increment_key_1_and_set_value_seq_0_1_and_emit), @bitSizeOf(Opcode));
+            try out.writeBits(entry.value.seq[0], 21);
+            try out.writeBits(entry.value.seq[1], 21);
+        } else if (seq_0_1_change and mem.eql(u8, &diff.key, &[4]u8{ 0, 0, 1, 0 })) {
+            try out.writeBits(@enumToInt(Opcode.increment_key_2_and_set_value_seq_0_1_and_emit), @bitSizeOf(Opcode));
+            try out.writeBits(entry.value.seq[0], 21);
+            try out.writeBits(entry.value.seq[1], 21);
+        } else {
+            // Within this category, of all diff.value.seq[0..] values:
+            // * 60% fit in 2 bits.
+            // * 84% fit in 8 bits.
+            var fits_2 = true;
+            var fits_8 = true;
+            for (diff.value.seq[0..entry.value.len]) |s| {
+                if (s >= (1 << 2)) {
+                    fits_2 = false;
+                }
+                if (s >= (1 << 8)) {
+                    fits_8 = false;
+                }
+            }
+
+            if (fits_2) {
+                try out.writeBits(@enumToInt(Opcode.set_key_and_value_seq_2bit_and_emit), @bitSizeOf(Opcode));
+                _ = try out.write(entry.key[0..entry.key_len]);
+                for (diff.value.seq[0..entry.value.len]) |s| try out.writeBits(s, 2);
+            } else if (fits_8) {
+                try out.writeBits(@enumToInt(Opcode.set_key_and_value_seq_8bit_and_emit), @bitSizeOf(Opcode));
+                _ = try out.write(entry.key[0..entry.key_len]);
+                for (diff.value.seq[0..entry.value.len]) |s| try out.writeBits(s, 8);
+            } else {
+                try out.writeBits(@enumToInt(Opcode.set_key_and_value_seq_21bit_and_emit), @bitSizeOf(Opcode));
+                _ = try out.write(entry.key[0..entry.key_len]);
+                for (diff.value.seq[0..entry.value.len]) |s| try out.writeBits(s, 21);
```

Comment:
Also note that many of these conditionals are just "more space efficient" variants of some simpler conditional, e.g. one can comment out the first two `if`s here and it would fallback to this 21 bit full encoding and work perfectly. The way I developed this was starting with fairly large encodings, evaluating how much space various opcodes took up, and then adding more space-efficient variants like the 2/8bit variants above here.